File: go/437.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func helper(t *TreeNode, target int64) int {
  if t == nil {
    return 0
  }
  target -= int64(t.Val)
  if target == 0 {
    return 1 + helper(t.Left, target) + helper(t.Right, target)
  }
  return helper(t.Left, target) + helper(t.Right, target)
}

func pathSum(root *TreeNode, targetSum int) int {
	var ret int = 0
	var cur *TreeNode = root
	var st []*TreeNode = make([]*TreeNode, 0)
	for cur != nil || len(st) > 0 {
		for cur != nil {
			st = append(st, cur)
			cur = cur.Left
		}

		cur = st[len(st)-1]
		st = st[:len(st)-1]
		ret  += helper(cur, int64(targetSum))

		cur = cur.Right
	}
	return ret
}

File: cpp/437.cpp
#include <iostream>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
    // check TreeNode t is valid or not
  int check(TreeNode *t, long target) {
    if (!t) // t is nullptr
      return 0;
    target -= t->val;
    if (!target) // target == 0
        return 1 + check(t->left, target) + check(t->right, target);
    return check(t->left, target) + check(t->right, target);
  }
  int pathSum(TreeNode *root, int targetSum) {
    int ret = 0;
    vector<TreeNode *> st;
    TreeNode *cur = root;
    // iteratively dfs, using one stack
    while (!st.empty() || cur != nullptr) {
      while (cur != nullptr) {
        st.push_back(cur);
        cur = cur->left;
      }
      cur = st.back();
      ret += check(cur, long(targetSum));
      st.pop_back();

      cur = cur->right;
    }
    return ret;
  }
};
